services/wfm/src/main/java/org/openkilda/wfm/topology/flow/service/FlowService.java

```diff
@@ -0,0 +1,377 @@
+
+            transactionManager.commit();
+
+            // To avoid race condition in DB updates, we should send commands only after DB transaction commit.
+            sender.sendInstallRulesCommand(flowPair.getForward(), forwardSegments);
+            sender.sendInstallRulesCommand(flowPair.getReverse(), reverseSegments);
+
+            return flowPair;
+        } catch (Exception e) {
+            transactionManager.rollback();
+            throw e;
+        }
+    }
+
+    private FlowPair buildFlowPair(Flow flow, PathPair pathPair) {
+        Instant timestamp = Instant.now();
+
+        Flow forward = flow.toBuilder()
+                .timeModify(timestamp)
+                .flowPath(pathPair.getForward())
+                .build();
+        Flow reverse = flow.toBuilder()
+                .timeModify(timestamp)
+                .srcSwitch(flow.getDestSwitch())
+                .srcPort(flow.getDestPort())
+                .srcVlan(flow.getDestVlan())
+                .destSwitch(flow.getSrcSwitch())
+                .destPort(flow.getSrcPort())
+                .destVlan(flow.getSrcVlan())
+                .flowPath(pathPair.getReverse())
+                .build();
+        return FlowPair.builder().forward(forward).reverse(reverse).build();
+    }
+
+    /**
+     * Stores a flow and related entities into DB, and invokes flow rules installation via the command sender.
+     *
+     * @param flowPair the flow to be saved.
+     * @param sender   the command sender for flow rules installation.
+     */
+    public void saveFlow(FlowPair flowPair, FlowCommandSender sender) {
+        transactionManager.begin();
+        try {
+            flowResourcesManager.registerUsedByFlow(flowPair);
+
+            flowRepository.createOrUpdate(flowPair);
+
+            final List<FlowSegment> forwardSegments = createFlowSegments(flowPair.getForward());
+            final List<FlowSegment> reverseSegments = createFlowSegments(flowPair.getReverse());
+
+            transactionManager.commit();
+
+            // To avoid race condition in DB updates, we should send commands only after DB transaction commit.
+            sender.sendInstallRulesCommand(flowPair.getForward(), forwardSegments);
+            sender.sendInstallRulesCommand(flowPair.getReverse(), reverseSegments);
+        } catch (Exception e) {
+            transactionManager.rollback();
+            throw e;
+        }
+    }
+
+    private List<FlowSegment> createFlowSegments(Flow flow) {
+        List<FlowSegment> segments = new ArrayList<>();
+
+        List<FlowPath.Node> nodes = flow.getFlowPath().getNodes();
+        for (int i = 0; i < nodes.size(); i += 2) {
+            FlowPath.Node src = nodes.get(i);
+            FlowPath.Node dst = nodes.get(i + 1);
+
+            FlowSegment flowSegment = FlowSegment.builder()
+                    .flowId(flow.getFlowId())
+                    .srcSwitch(Switch.builder().switchId(src.getSwitchId()).build())
```

Comment:
This can be treated incorrectly by OGM. We should try to use valid/existing switch objects to define relations.

Reply:
OGM requires only IDs of relation's start and end node, not complete entities.  As soon as Neo4jFlowRepository cares about mapping by ID for switch entities, OGM receives Flow with properly obtained and set Switch entities.